File: XLPagerTabStrip/XL/Controllers/XLPagerTabStripViewController.m
//
//  XLPagerTabStripViewController
//  XLPagerTabStrip ( https://github.com/xmartlabs/XLPagerTabStrip )
//
//  Copyright (c) 2015 Xmartlabs ( http://xmartlabs.com )
//
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "XLPagerTabStripViewController.h"

@interface XLPagerTabStripViewController ()

@property (nonatomic) NSUInteger currentIndex;

@end

@implementation XLPagerTabStripViewController
{
    NSUInteger _lastPageNumber;
    CGFloat _lastContentOffset;
    NSUInteger _pageBeforeRotate;
    NSArray * _originalPagerTabStripChildViewControllers;
    CGSize _lastSize;
}

@synthesize currentIndex = _currentIndex;

#pragma maek - initializers

-(id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self){
        [self pagerTabStripViewControllerInit];
    }
    return self;
}


-(id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self){
        [self pagerTabStripViewControllerInit];
    }
    return self;
}

-(void)dealloc
{
    self.containerView.delegate = nil;
}


-(void)pagerTabStripViewControllerInit
{
    _currentIndex = 0;
    _delegate = self;
    _dataSource = self;
    _lastContentOffset = 0.0f;
    _isElasticIndicatorLimit = NO;
    _skipIntermediateViewControllers = YES;
    _isProgressiveIndicator = NO;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    if (!self.containerView){
        self.containerView = [[UIScrollView alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.view.bounds))];
        self.containerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;

        [self.view addSubview:self.containerView];
    }
    self.containerView.bounces = YES;
    [self.containerView setAlwaysBounceHorizontal:YES];
    [self.containerView setAlwaysBounceVertical:NO];
    self.containerView.scrollsToTop = NO;
    self.containerView.delegate = self;
    self.containerView.showsVerticalScrollIndicator = NO;
    self.containerView.showsHorizontalScrollIndicator = NO;
    self.containerView.pagingEnabled = YES;
    
    if (self.dataSource){
        _pagerTabStripChildViewControllers = [self.dataSource childViewControllersForPagerTabStripViewController:self];
    }
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
}

-(void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    _lastSize = self.containerView.bounds.size;
    [self updateIfNeeded];
}

-(void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    [self updateIfNeeded];
    if  ([[[UIDevice currentDevice] systemVersion] compare:@"8.0" options:NSNumericSearch] == NSOrderedAscending){
        // SYSTEM_VERSION_LESS_THAN 8.0
        [self.view layoutSubviews];
    }
}

#pragma mark - move to another view controller

-(void)moveToViewControllerAtIndex:(NSUInteger)index
{
    [self moveToViewControllerAtIndex:index animated:YES];
}


-(void)moveToViewControllerAtIndex:(NSUInteger)index animated:(BOOL)animated
{
    if (!self.isViewLoaded || !self.view.window){
        self.currentIndex = index;
    }
    else{
        if (animated && self.skipIntermediateViewControllers && ABS(self.currentIndex - index) > 1){
            NSArray * originalPagerTabStripChildViewControllers = self.pagerTabStripChildViewControllers;
            NSMutableArray * tempChildViewControllers = [NSMutableArray arrayWithArray:originalPagerTabStripChildViewControllers];
            UIViewController * currentChildVC = [originalPagerTabStripChildViewControllers objectAtIndex:self.currentIndex];
            NSUInteger fromIndex = (self.currentIndex < index) ? index - 1 : index + 1;
            [tempChildViewControllers setObject:[originalPagerTabStripChildViewControllers objectAtIndex:fromIndex] atIndexedSubscript:self.currentIndex];
            [tempChildViewControllers setObject:currentChildVC atIndexedSubscript:fromIndex];
            _pagerTabStripChildViewControllers = tempChildViewControllers;
            [self.containerView setContentOffset:CGPointMake([self pageOffsetForChildIndex:fromIndex], 0) animated:NO];
            if (self.navigationController){
                self.navigationController.view.userInteractionEnabled = NO;
            }
            else{
                self.view.userInteractionEnabled = NO;
            }
            _originalPagerTabStripChildViewControllers = originalPagerTabStripChildViewControllers;
            [self.containerView setContentOffset:CGPointMake([self pageOffsetForChildIndex:index], 0) animated:YES];
        }
        else{
            [self.containerView setContentOffset:CGPointMake([self pageOffsetForChildIndex:index], 0) animated:animated];
        }
        
    }
}


-(void)moveToViewController:(UIViewController *)viewController
{
    [self moveToViewControllerAtIndex:[self.pagerTabStripChildViewControllers indexOfObject:viewController]];
}

-(void)moveToViewController:(UIViewController *)viewController animated:(BOOL)animated {
    [self moveToViewControllerAtIndex:[self.pagerTabStripChildViewControllers indexOfObject:viewController] animated:animated];
}

#pragma mark - XLPagerTabStripViewControllerDelegate

-(void)pagerTabStripViewController:(XLPagerTabStripViewController *)pagerTabStripViewController
          updateIndicatorFromIndex:(NSInteger)fromIndex
                           toIndex:(NSInteger)toIndex{

}

-(void)pagerTabStripViewController:(XLPagerTabStripViewController *)pagerTabStripViewController
          updateIndicatorFromIndex:(NSInteger)fromIndex
                           toIndex:(NSInteger)toIndex
            withProgressPercentage:(CGFloat)progressPercentage
             indexWasChanged:(BOOL)indexWasChanged
{
}


#pragma mark - XLPagerTabStripViewControllerDataSource

-(NSArray *)childViewControllersForPagerTabStripViewController:(XLPagerTabStripViewController *)pagerTabStripViewController
{
    return self.pagerTabStripChildViewControllers;
}


#pragma mark - Helpers

-(void)updateIfNeeded
{
    if (!CGSizeEqualToSize(_lastSize, self.containerView.bounds.size)){
        [self updateContent];
    }
}

-(XLPagerTabStripDirection)scrollDirection
{
    if (self.containerView.contentOffset.x > _lastContentOffset){
        return XLPagerTabStripDirectionLeft;
    }
    else if (self.containerView.contentOffset.x < _lastContentOffset){
        return XLPagerTabStripDirectionRight;
    }
    return XLPagerTabStripDirectionNone;
}

-(BOOL)canMoveToIndex:(NSUInteger)index
{
    return (self.currentIndex != index && self.pagerTabStripChildViewControllers.count > index);
}

-(CGFloat)pageOffsetForChildIndex:(NSUInteger)index
{
    return (index * CGRectGetWidth(self.containerView.bounds));
}

-(CGFloat)offsetForChildIndex:(NSUInteger)index
{
    return (index * CGRectGetWidth(self.containerView.bounds) + ((CGRectGetWidth(self.containerView.bounds) - CGRectGetWidth(self.view.bounds)) * 0.5));
}

-(CGFloat)offsetForChildViewController:(UIViewController *)viewController
{
    NSInteger index = [self.pagerTabStripChildViewControllers indexOfObject:viewController];
    if (index == NSNotFound){
        @throw [NSException exceptionWithName:NSRangeException reason:nil userInfo:nil];
    }
    return [self offsetForChildIndex:index];
}

-(NSUInteger)pageForContentOffset:(CGFloat)contentOffset
{
    NSInteger result = [self virtualPageForContentOffset:contentOffset];
    return [self pageForVirtualPage:result];
}

-(NSInteger)virtualPageForContentOffset:(CGFloat)contentOffset
{
    NSInteger result = (contentOffset + (1.5f * [self pageWidth])) / [self pageWidth];
    return result - 1;
}

-(NSUInteger)pageForVirtualPage:(NSInteger)virtualPage
{
    if (virtualPage < 0){
        return 0;
    }
    if (virtualPage > self.pagerTabStripChildViewControllers.count - 1){
        return self.pagerTabStripChildViewControllers.count - 1;
    }
    return virtualPage;
}

-(CGFloat)pageWidth
{
    return CGRectGetWidth(self.containerView.bounds);
}

-(CGFloat)scrollPercentage
{
    if ([self scrollDirection] == XLPagerTabStripDirectionLeft || [self scrollDirection] == XLPagerTabStripDirectionNone){
        if (fmodf(self.containerView.contentOffset.x, [self pageWidth]) == 0.0) {
            return 1.0;
        }
        return fmodf(self.containerView.contentOffset.x, [self pageWidth]) / [self pageWidth];
    }
    return 1 - fmodf(self.containerView.contentOffset.x >= 0 ? self.containerView.contentOffset.x : [self pageWidth] + self.containerView.contentOffset.x, [self pageWidth]) / [self pageWidth];
}

-(void)updateContent
{
    if (!CGSizeEqualToSize(_lastSize, self.containerView.bounds.size)){
        if (_lastSize.width != self.containerView.bounds.size.width){
            _lastSize = self.containerView.bounds.size;
            [self.containerView setContentOffset:CGPointMake([self pageOffsetForChildIndex:self.currentIndex], 0) animated:NO];
        }
        else{
            _lastSize = self.containerView.bounds.size;
        }
    }
    NSArray * childViewControllers = self.pagerTabStripChildViewControllers;
    self.containerView.contentSize = CGSizeMake(CGRectGetWidth(self.containerView.bounds) * childViewControllers.count, self.containerView.contentSize.height);
    
    [childViewControllers enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        UIViewController * childController = (UIViewController *)obj;
        CGFloat pageOffsetForChild = [self pageOffsetForChildIndex:idx];
        if (fabs(self.containerView.contentOffset.x - pageOffsetForChild) < CGRectGetWidth(self.containerView.bounds)) {
            if (![childController parentViewController]) { // Add child
                [childController beginAppearanceTransition:YES animated:NO];
                [self addChildViewController:childController];
                
                CGFloat childPosition = [self offsetForChildIndex:idx];
                [childController.view setFrame:CGRectMake(childPosition, 0, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.containerView.bounds))];
                childController.view.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
                
                [self.containerView addSubview:childController.view];
                [childController didMoveToParentViewController:self];
                [childController endAppearanceTransition];
            } else {
                CGFloat childPosition = [self offsetForChildIndex:idx];
                [childController.view setFrame:CGRectMake(childPosition, 0, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.containerView.bounds))];
                childController.view.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
            }
        } else {
            if ([childController parentViewController]) { // Remove child
                [childController willMoveToParentViewController:nil];
                [childController beginAppearanceTransition:NO animated:NO];
                [childController.view removeFromSuperview];
                [childController removeFromParentViewController];
                [childController endAppearanceTransition];
            }
        }
    }];
    
    NSUInteger oldCurrentIndex = self.currentIndex;
    NSInteger virtualPage = [self virtualPageForContentOffset:self.containerView.contentOffset.x];
    NSUInteger newCurrentIndex = [self pageForVirtualPage:virtualPage];
    self.currentIndex = newCurrentIndex;
    BOOL changeCurrentIndex = newCurrentIndex != oldCurrentIndex;
    
    if (self.isProgressiveIndicator){
        if ([self.delegate respondsToSelector:@selector(pagerTabStripViewController:updateIndicatorFromIndex:toIndex:withProgressPercentage:indexWasChanged:)]){
            CGFloat scrollPercentage = [self scrollPercentage];
            if (scrollPercentage > 0) {
                NSInteger fromIndex = self.currentIndex;
                NSInteger toIndex = self.currentIndex;
                XLPagerTabStripDirection scrollDirection = [self scrollDirection];
                if (scrollDirection == XLPagerTabStripDirectionLeft){
                    if (virtualPage > self.pagerTabStripChildViewControllers.count - 1){
                        fromIndex = self.pagerTabStripChildViewControllers.count - 1;
                        toIndex = self.pagerTabStripChildViewControllers.count;
                    }
                    else{
                        if (scrollPercentage >= 0.5f){
                            fromIndex = MAX(toIndex - 1, 0);
                        }
                        else{
                            toIndex = fromIndex + 1;
                        }
                    }
                }
                else if (scrollDirection == XLPagerTabStripDirectionRight) {
                    if (virtualPage < 0){
                        fromIndex = 0;
                        toIndex = -1;
                    }
                    else{
                        if (scrollPercentage > 0.5f){
                            fromIndex = MIN(toIndex + 1, self.pagerTabStripChildViewControllers.count - 1);
                        }
                        else{
                            toIndex = fromIndex - 1;
                        }
                    }
                }
                [self.delegate pagerTabStripViewController:self updateIndicatorFromIndex:fromIndex toIndex:toIndex withProgressPercentage:(self.isElasticIndicatorLimit ? scrollPercentage : ( toIndex < 0 || toIndex >= self.pagerTabStripChildViewControllers.count ? 0 : scrollPercentage )) indexWasChanged:changeCurrentIndex];
            }
        }
    }
    else{
        if ([self.delegate respondsToSelector:@selector(pagerTabStripViewController:updateIndicatorFromIndex:toIndex:)] && oldCurrentIndex != newCurrentIndex){
            [self.delegate pagerTabStripViewController:self
                              updateIndicatorFromIndex:MIN(oldCurrentIndex, self.pagerTabStripChildViewControllers.count - 1)
                                               toIndex:newCurrentIndex];
        }
    }
}


-(void)reloadPagerTabStripView
{
    if ([self isViewLoaded]){
        [self.pagerTabStripChildViewControllers enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
            UIViewController * childController = (UIViewController *)obj;
            if ([childController parentViewController]){
                [childController.view removeFromSuperview];
                [childController willMoveToParentViewController:nil];
                [childController removeFromParentViewController];
            }
        }];
        _pagerTabStripChildViewControllers = self.dataSource ? [self.dataSource childViewControllersForPagerTabStripViewController:self] : @[];
        self.containerView.contentSize = CGSizeMake(CGRectGetWidth(self.containerView.bounds) * _pagerTabStripChildViewControllers.count, self.containerView.contentSize.height);
        if (self.currentIndex >= _pagerTabStripChildViewControllers.count){
            self.currentIndex = _pagerTabStripChildViewControllers.count - 1;
        }
        [self.containerView setContentOffset:CGPointMake([self pageOffsetForChildIndex:self.currentIndex], 0)  animated:NO];
        [self updateContent];
    }
}

#pragma mark - UIScrollViewDelegte

-(void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    if (self.containerView == scrollView){
        [self updateContent];
    }
}


-(void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
{
    if (self.containerView == scrollView){
        _lastPageNumber = [self pageForContentOffset:scrollView.contentOffset.x];
        _lastContentOffset = scrollView.contentOffset.x;
    }
}

-(void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView
{
    if (self.containerView == scrollView && _originalPagerTabStripChildViewControllers){
        _pagerTabStripChildViewControllers = _originalPagerTabStripChildViewControllers;
        _originalPagerTabStripChildViewControllers = nil;
        if (self.navigationController){
            self.navigationController.view.userInteractionEnabled = YES;
        }
        else{
            self.view.userInteractionEnabled = YES;
        }
        [self updateContent];
    }
}



#pragma mark - Orientation

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
{
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    _pageBeforeRotate = self.currentIndex;
        __typeof__(self) __weak weakSelf = self;
    
    UIInterfaceOrientation fromOrientation = [[UIApplication sharedApplication] statusBarOrientation];
    [coordinator animateAlongsideTransition:nil
                                 completion:^(id<UIViewControllerTransitionCoordinatorContext> context) {
        [weakSelf didRotateFromInterfaceOrientation:fromOrientation];
    }];
}

-(void)willRotateToInterfaceOrientation:(UIInterfaceOrientation)toInterfaceOrientation duration:(NSTimeInterval)duration
{
    _pageBeforeRotate = self.currentIndex;
}

-(void)didRotateFromInterfaceOrientation:(UIInterfaceOrientation)fromInterfaceOrientation
{
    self.currentIndex = _pageBeforeRotate;
    [self updateIfNeeded];
}


@end

File: Sources/XLPagerTabStripViewController.swift
//
//  XLPagerTabStripViewController.swift
//  XLPagerTabStrip
//
//  Created by Santiago on 12/30/15.
//  Copyright © 2015 Xmartlabs. All rights reserved.
//

import Foundation

public struct ChildHeader {
    
    public var title: String
    public var image: UIImage?
    public var highlightedImage: UIImage?
    public var color: UIColor?
    
    public init(title: String) {
        self.title = title
    }
    
    public init(title: String, image: UIImage?) {
        self.init(title: title)
        self.image = image
    }
    
    public init(title: String, image: UIImage?, highlightedImage: UIImage?) {
        self.init(title: title, image: image)
        self.highlightedImage = highlightedImage
    }
    
    public init(title: String, image: UIImage?, highlightedImage: UIImage?, color: UIColor?){
        self.init(title: title, image: image, highlightedImage: highlightedImage)
        self.color = color
    }
}

public protocol XLPagerTabStripChildItem{
    func childHeaderForPagerTabStripViewController(pagerTabStripController: XLPagerTabStripViewController) -> ChildHeader
}

public enum XLPagerTabStripDirection{
    case Left
    case Right
    case None
}

public protocol XLPagerTabStripViewControllerDelegate{
    func pagerTabStripViewController(pagerTabStripViewController: XLPagerTabStripViewController, updateIndicatorFromIndex fromIndex: Int, toIndex index: Int)
    func pagerTabStripViewController(pagerTabStripViewController: XLPagerTabStripViewController, updateIndicatorFromIndex fromIndex: Int, toIndex index: Int, withProgressPercentage progressPercentage: Float, indexWasChanged changed: Bool)
}

public protocol XLPagerTabStripViewControllerDataSource{
    func childViewControllersForPagerTabStripViewController(pagerTabStripController: XLPagerTabStripViewController) ->[UIViewController]?
}

public class XLPagerTabStripViewController: UIViewController, UIScrollViewDelegate, XLPagerTabStripViewControllerDataSource, XLPagerTabStripViewControllerDelegate{
    private(set) var pagerTabStripChildViewControllers : [UIViewController]?
    public var containerView: UIScrollView?
    public var delegate: XLPagerTabStripViewControllerDelegate?
    public var datasource: XLPagerTabStripViewControllerDataSource?

    
    private(set) var currentIndex: Int = 0
    public var skipIntermediateViewControllers: Bool = false
    public var isProgressiveIndicator: Bool = false
    public var isElasticIndicatorLimit: Bool = false
    
    private var pagerTabStripChildViewControllersForScrolling : [UIViewController]?
    
    private var lastPageNumber: Int = 0
    private var lastContentOffset: Float = 0.0
    private var pageBeforeRotate: Int = 0
    private var lastSize: CGSize = CGSizeMake(0, 0)
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?){
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        self.pagerTabStripViewControllerInit()
    }

    required public init?(coder aDecoder: NSCoder){
        super.init(coder: aDecoder)
        self.pagerTabStripViewControllerInit()
    }
    
    func pagerTabStripViewControllerInit(){
        self.delegate = self;
        self.datasource = self;
    }
    
    override public func viewDidLoad(){
        super.viewDidLoad()
        
        if (self.containerView == nil){
            self.containerView = UIScrollView(frame: CGRectMake(0, 0, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.view.bounds)))
            self.containerView?.autoresizingMask = [.FlexibleWidth, .FlexibleHeight]
            self.view.addSubview(self.containerView!)
        }
        
        self.containerView?.bounces = true
        self.containerView?.alwaysBounceHorizontal = true
        self.containerView?.alwaysBounceVertical = false
        self.containerView?.scrollsToTop = false
        self.containerView?.delegate = self
        self.containerView?.showsVerticalScrollIndicator = false
        self.containerView?.showsHorizontalScrollIndicator = false
        self.containerView?.pagingEnabled = true
        
        self.pagerTabStripChildViewControllers = self.datasource?.childViewControllersForPagerTabStripViewController(self)
    }
    
    override public func viewDidAppear(animated: Bool){
        super.viewDidAppear(animated)
        self.lastSize = self.containerView!.bounds.size
        self.updateIfNeeded()
    }
    
    override public func viewDidLayoutSubviews(){
        super.viewDidLayoutSubviews()
        self.updateIfNeeded()
        switch UIDevice.currentDevice().systemVersion.compare("8.0", options: NSStringCompareOptions.NumericSearch){
            case .OrderedAscending:
                self.view.layoutSubviews()
            default: break
        }

    }
    
    public func moveToViewControllerAtIndex(index: Int){
        self.moveToViewControllerAtIndex(index, animated: true)
    }
    
    public func moveToViewControllerAtIndex(index: Int, animated: Bool){
        if (!self.isViewLoaded() || self.view.window == nil){
            self.currentIndex = index
        }
        else{
            if (animated && self.skipIntermediateViewControllers && abs(self.currentIndex - index) > 1){
                var tmpChildViewControllers = self.pagerTabStripChildViewControllers
                let currentChildVC = self.pagerTabStripChildViewControllers![self.currentIndex]
                let fromIndex = self.currentIndex < index ? index - 1 :index + 1
                let fromChildVC = self.pagerTabStripChildViewControllers![fromIndex]
                tmpChildViewControllers![self.currentIndex] = fromChildVC
                tmpChildViewControllers![fromIndex] = currentChildVC
                pagerTabStripChildViewControllersForScrolling = tmpChildViewControllers
                self.containerView?.contentOffset = CGPointMake(self.pageOffsetForChildIndex(index: fromIndex), 0)
                if((self.navigationController) != nil){
                    self.navigationController?.view.userInteractionEnabled = false;
                }
                else{
                    self.view.userInteractionEnabled = false
                }
                self.containerView?.contentOffset = CGPointMake(self.pageOffsetForChildIndex(index: index), 0)
            }
            else{
                self.containerView?.setContentOffset(CGPointMake(self.pageOffsetForChildIndex(index: index), 0), animated: true)
            }
        }
    }
    
    public func moveToViewController(viewController: UIViewController){
        self.moveToViewControllerAtIndex(self.pagerTabStripChildViewControllers!.indexOf(viewController)!)
    }

    public func moveToViewController(viewController: UIViewController, animated: Bool){
        self.moveToViewControllerAtIndex(self.pagerTabStripChildViewControllers!.indexOf(viewController)!, animated: animated)
    }
    
    //MARK: - XLPagerTabStripViewControllerDataSource
    public func childViewControllersForPagerTabStripViewController(pagerTabStripController: XLPagerTabStripViewController) -> [UIViewController]?{
        assert(false, "Sub-class must implement the XLPagerTabStripViewControllerDataSource childViewControllersForPagerTabStripViewController: method")
        return nil
    }
    
    //MARK: - XLPagerTabStripViewControllerDelegate
    public func pagerTabStripViewController(pagerTabStripViewController: XLPagerTabStripViewController, updateIndicatorFromIndex fromIndex: Int, toIndex index: Int){
        
    }
    
    public func pagerTabStripViewController(pagerTabStripViewController: XLPagerTabStripViewController, updateIndicatorFromIndex fromIndex: Int, toIndex index: Int, withProgressPercentage progressPercentage: Float, indexWasChanged changed: Bool){
        
    }
    
    //MARK: - Helpers
    func getPagerTabStripChildViewControllersForScrolling() -> [UIViewController]?{
        if let controllers = self.pagerTabStripChildViewControllersForScrolling{
            return controllers
        }
        
        return pagerTabStripChildViewControllers!
    }
    
    public func updateIfNeeded(){
        if (!CGSizeEqualToSize(self.lastSize, self.containerView!.bounds.size)){
            self.updateContent()
        }
    }
    
    public func scrollDirection() -> XLPagerTabStripDirection{
        if (Float(self.containerView!.contentOffset.x) > self.lastContentOffset){
            return XLPagerTabStripDirection.Left
        }
        else if (Float(self.containerView!.contentOffset.x) < self.lastContentOffset){
            return XLPagerTabStripDirection.Right
        }
        
        return XLPagerTabStripDirection.None
    }
    
    public func canMoveToIndex(index index: Int) -> Bool{
        return (self.currentIndex != index && self.pagerTabStripChildViewControllers?.count > index)
    }

    public func pageOffsetForChildIndex(index index: Int) -> CGFloat{
        return (CGFloat(index) * CGRectGetWidth(self.containerView!.bounds))
    }
    
    public func offsetForChildIndex(index index: Int) -> CGFloat{
        return (CGFloat(index) * CGRectGetWidth(self.containerView!.bounds) + ((CGRectGetWidth(self.containerView!.bounds) - CGRectGetWidth(self.view.bounds)) * 0.5))
    }
    
    enum Exception : ErrorType {
        case RangeException(String)
    }
    
    public func offsetForChildViewController(viewController: UIViewController) throws -> CGFloat{
        if let index = self.pagerTabStripChildViewControllers?.indexOf(viewController){
            return self.offsetForChildIndex(index: index)
        }
        
        throw Exception.RangeException(NSRangeException)
    }
    
    public func pageForContentOffset(contentOffset contentOffset: Float) -> Int{
        let result = self.virtualPageForContentOffset(contentOffset: contentOffset)
        return self.pageForVirtualPage(virtualPage: result)
    }
    
    public func virtualPageForContentOffset(contentOffset contentOffset: Float) -> Int{
        let result: Int = Int((contentOffset + (1.5 * self.pageWidth())) / self.pageWidth())
        return result - 1
    }
    
    public func pageForVirtualPage(virtualPage virtualPage: Int) -> Int{
        if (virtualPage < 0){
            return 0
        }
        if (virtualPage > self.pagerTabStripChildViewControllers!.count - 1){
            return self.pagerTabStripChildViewControllers!.count - 1
        }
        return virtualPage
    }
    
    public func pageWidth() -> Float{
        return Float(CGRectGetWidth(self.containerView!.bounds))
    }
    
    public func scrollPercentage() -> Float{
        if (self.scrollDirection() == .Left || self.scrollDirection() == .None){
            if (fmodf(Float(self.containerView!.contentOffset.x), self.pageWidth()) == 0.0){
                return 1.0
            }
            return fmodf(Float(self.containerView!.contentOffset.x), self.pageWidth() / self.pageWidth())
        }
        return 1 - fmodf(self.containerView!.contentOffset.x >= 0 ? Float(self.containerView!.contentOffset.x) : self.pageWidth() + Float(self.containerView!.contentOffset.x), self.pageWidth()) / self.pageWidth();
    }
    
    public func updateContent(){
        if (!CGSizeEqualToSize(self.lastSize, self.containerView!.bounds.size)){
            if (self.lastSize.width != self.containerView!.bounds.size.width){
                self.lastSize = self.containerView!.bounds.size
                self.containerView?.contentOffset = CGPointMake(self.pageOffsetForChildIndex(index: self.currentIndex), 0)
            }
            else{
                self.lastSize = self.containerView!.bounds.size
            }
        }
        
        let childViewControllers = self.getPagerTabStripChildViewControllersForScrolling()
        self.containerView?.contentSize = CGSizeMake(CGRectGetWidth(self.containerView!.bounds) * CGFloat(childViewControllers!.count), self.containerView!.contentSize.height)
        
        for (index, childController) in childViewControllers!.enumerate(){
            let pageOffsetForChild = self.pageOffsetForChildIndex(index: index)
            if (fabs(self.containerView!.contentOffset.x - pageOffsetForChild) < CGRectGetWidth(self.containerView!.bounds)){
                if (childController.parentViewController == nil){
                    self.addChildViewController(childController)
                    childController.beginAppearanceTransition(true, animated: false)
                    
                    let childPosition = self.offsetForChildIndex(index: index)
                    childController.view.frame = CGRectMake(childPosition, 0, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.containerView!.bounds))
                    childController.view.autoresizingMask = [.FlexibleHeight, .FlexibleWidth]
                    
                    self.containerView?.addSubview(childController.view)
                    childController.didMoveToParentViewController(self)
                    childController.endAppearanceTransition()
                }
                else{
                    let childPosition = self.offsetForChildIndex(index: index)
                    childController.view.frame = CGRectMake(childPosition, 0, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.containerView!.bounds))
                    childController.view.autoresizingMask = [.FlexibleHeight, .FlexibleWidth]
                    
                }
            }
            else{
                if (childController.parentViewController != nil){
                    childController.willMoveToParentViewController(nil)
                    childController.beginAppearanceTransition(false, animated: false)
                    childController.view.removeFromSuperview()
                    childController.removeFromParentViewController()
                    childController.endAppearanceTransition()
                }
            }
        }
        
        let oldCurrentIndex = self.currentIndex
        let virtualPage = self.virtualPageForContentOffset(contentOffset: Float(self.containerView!.contentOffset.x))
        let newCurrentIndex = self.pageForVirtualPage(virtualPage: virtualPage)
        self.currentIndex = newCurrentIndex
        let changeCurrentIndex = newCurrentIndex != oldCurrentIndex
        
        if (self.isProgressiveIndicator){
            // FIXME: - check if delegate implements? pagerTabStripViewController(pagerTabStripViewController: XLPagerTabStripViewController, updateIndicatorFromIndex fromIndex: Int, toIndex index: Int, withProgressPercentage progressPercentage: Float, indexWasChanged changed: Bool)
            
            let scrollPercentage = self.scrollPercentage()
            if (scrollPercentage > 0){
                var fromIndex = self.currentIndex
                var toIndex = self.currentIndex
                let scrollDirection = self.scrollDirection()
                
                if (scrollDirection == .Left){
                    if (virtualPage > self.getPagerTabStripChildViewControllersForScrolling()!.count - 1){
                        fromIndex = self.getPagerTabStripChildViewControllersForScrolling()!.count - 1
                        toIndex = self.getPagerTabStripChildViewControllersForScrolling()!.count
                    }
                    else{
                        if (scrollPercentage >= 0.5){
                            fromIndex = max(toIndex - 1, 0)
                        }
                        else{
                            toIndex = fromIndex + 1
                        }
                    }
                }
                else if (scrollDirection == .Right){
                    if (virtualPage < 0){
                        fromIndex = 0
                        toIndex = -1
                    }
                    else{
                        if (scrollPercentage > 0.5){
                            fromIndex = min(toIndex + 1, self.getPagerTabStripChildViewControllersForScrolling()!.count - 1)
                        }
                        else{
                            toIndex = fromIndex - 1
                        }
                    }
                }
                self.delegate?.pagerTabStripViewController(self, updateIndicatorFromIndex: fromIndex, toIndex: toIndex, withProgressPercentage: (self.isElasticIndicatorLimit ? scrollPercentage : (toIndex < 0 || toIndex >= self.getPagerTabStripChildViewControllersForScrolling()!.count ? 0 : scrollPercentage)), indexWasChanged: changeCurrentIndex)
            }
        }
        else{
            self.delegate?.pagerTabStripViewController(self, updateIndicatorFromIndex: min(oldCurrentIndex, self.getPagerTabStripChildViewControllersForScrolling()!.count - 1), toIndex: newCurrentIndex)
        }
    }
    
    public func reloadPagerTabStripView()
    {
        if (self.isViewLoaded()){
            for item in self.pagerTabStripChildViewControllers!.enumerate(){
                let childController = item.element
                if (childController.parentViewController != nil){
                    childController.view.removeFromSuperview()
                    childController.willMoveToParentViewController(nil)
                    childController.removeFromParentViewController()
                }
            }
            
            pagerTabStripChildViewControllers = (self.datasource != nil) ? self.datasource?.childViewControllersForPagerTabStripViewController(self) : []
            self.containerView?.contentSize = CGSizeMake(CGRectGetWidth(self.containerView!.bounds) * CGFloat(self.pagerTabStripChildViewControllers!.count), self.containerView!.contentSize.height)
            if (currentIndex >= self.pagerTabStripChildViewControllers!.count){
                currentIndex = pagerTabStripChildViewControllers!.count - 1
            }
            self.containerView?.contentOffset = CGPointMake(self.pageOffsetForChildIndex(index: currentIndex), 0)
            self.updateContent()
        }
    }
    
    //MARK: - UIScrollDelegate
    
    public func scrollViewDidScroll(scrollView: UIScrollView) {
        if (self.containerView == scrollView){
            self.updateContent()
        }
    }
    
    public func scrollViewWillBeginDragging(scrollView: UIScrollView) {
        if (self.containerView == scrollView){
            lastPageNumber = self.pageForContentOffset(contentOffset: Float(scrollView.contentOffset.x))
            lastContentOffset = Float(scrollView.contentOffset.x)
        }
    }
    
    public func scrollViewDidEndScrollingAnimation(scrollView: UIScrollView) {
        if (self.containerView == scrollView && getPagerTabStripChildViewControllersForScrolling() != nil){
            pagerTabStripChildViewControllersForScrolling = nil
            if (navigationController != nil){
                self.navigationController?.view.userInteractionEnabled = true
            }
            else{
                self.view.userInteractionEnabled = true
            }
            self.updateContent()
        }
    }
    
    //MARK: - Orientation
    
    public override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransitionToSize(size, withTransitionCoordinator: coordinator)
        pageBeforeRotate = currentIndex
        coordinator.animateAlongsideTransition(nil) { (context) -> Void in
            self.currentIndex = self.pageBeforeRotate
            self.updateIfNeeded()
        }
    }
    
    public override func willRotateToInterfaceOrientation(toInterfaceOrientation: UIInterfaceOrientation, duration: NSTimeInterval) {
        pageBeforeRotate = currentIndex
    }
}
